homeassistant/components/crownstone/helpers.py

```diff
@@ -0,0 +1,31 @@
+"""Helper functions/classes for Crownstone."""
+import asyncio
+import threading
+
+from crownstone_uart import CrownstoneUart
+
+
+class UartManager(threading.Thread):
+    """Uart manager that manages usb connections."""
+
+    def __init__(self) -> None:
+        """Init with new event loop and instance."""
+        self.loop = asyncio.new_event_loop()
+        self.uart_instance = CrownstoneUart(self.loop)
```

Comment:
If this is an async package, why are you running it in a new event loop?

Reply:
The Uart lib is created by the Crownstone core members, with the idea to have it block until a USB is connected, so it works in a recursive way instead of creating tasks in the event loop to allow for other coroutines to be executed. If no USB is connected, it just keeps on scanning for USB dongles until it finds one. This way it would mean the USB could be connected and removed at any time, and the `async_turn_on` uses the the USB whenever it's available, and falls back to the cloud if it is not.

We've decided to have it run in a separate thread to stop it from blocking the event loop.

I can request changes, if you deem it necessary.

Reply:
This will actually try to connect to every serial port it can find, to see if do a successful Crowdstone handshake.

<https://github.com/crownstone/crownstone-lib-python-uart/blob/master/crownstone_uart/core/uart/UartManager.py#L90>

This is very dangerous and can interfere with other equipment in use. Since this type of communication is not uncommon, I would say this is not a good method to have in Home Assistant.

Reply:
Holy moly. Yeah that is 100% not allowed in HA. USB port need to be passed in.

Reply:
It is possible for the lib to take a port parameter and skip the handshake section. How do we ensure to provide the correct port though? Let's say I have the Crownstone dongle at /dev/ttyUSB0, but then i add an other USB dongle of some kind and restart HA, and the Crownstone dongle is now at /dev/ttyUSB1

Reply:
By not using those ports. We advise our users to use the linked `by-id` paths in general to avoid those issues.

Reply:
Another approach is to use a udev rule.  In my case, I have the following line in a file in the folder `/etc/udev/rules.d/`:

```
SUBSYSTEM=="tty", ATTRS{idVendor}=="10c4", ATTRS{idProduct}=="ea60", SYMLINK+="crownstone"
```

This makes the usb dongle available at port `/dev/crownstone`

Reply:
@hillstub That is a modification done on the system administration level. From a Python application perspective, that is not an option or requirement we can take into account. Let's keep this PR review on the topic of reviewing code at hand. Thanks 👍